### ISSUE TITLE
Override compatibility of complex types

### ISSUE THREAD

**Issue by Nnonexistent** [python/mypy#14929]
**Bug Report**

While checking for override compatibility on subclass methods, mypy incorrectly throws an error about incompatible return types, even if the type is a subset of base type.

Issue only occurs in complex types such as lists or dicts.

E.g. if original type was `list[int | str]` and subclass type is `list[str]` mypy will throw an error `[override]`.

**To Reproduce**

```python
class A:
    def func(self) -> list[int | str]:
        return [1]


class B(A):
    def func(self) -> list[str]:  # <- override error
        return ['qwe']
```
[Playground with examples](https://mypy-play.net/?mypy=latest&python=3.11&gist=31ee336a6014af9c02d122575003d6e9)

**Expected Behavior**

No errors should be reported as long as types are compatible.
 
**Actual Behavior**

```
error: Return type "List[str]" of "func" incompatible with return type "List[Union[int, str]]" in supertype "A"  [override]
```

**Your Environment**

- Mypy version used: 1.1.1
- Python version used: 3.11

**Comment by erictraut** [comment]
I think mypy's behavior is correct here. An override must be type compatible with the method it is overriding, and `list[str]` is not compatible with `list[int | str]`. The type parameter for `list` is invariant.

**Comment by Nnonexistent** [comment]
@erictraut Override is compatible: `list[str]` is a subset of `list[str | int]`. So whatever  subclass function will return it will be compatible with broader definition of a base function.

Also, without lists subset types works perfectly fine:

```python
class A:
    def func(self) -> int | str:
        return 1


class B(A):
    def func(self) -> str:  # <- no error here
        return 'qwe'
```

Also, please, check mentioned playground with more examples: https://mypy-play.net/?mypy=latest&python=3.11&gist=31ee336a6014af9c02d122575003d6e9

**Comment by erictraut** [comment]
> `list[str]` is a subset of `list[str | int]`

No, `list[str]` is not a subtype of `list[str | int]`. See [this documentation](https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types) for more details.
